### ISSUE TITLE
BinaryWeight Network

### ISSUE THREAD

**Issue by lusensama** [jiecaoyu/XNOR-Net-PyTorch#78]
Hello,

In the original paper and code, there is a part of BinaryWeight Network.
In fact, the original XNORNET in torch provides an option to train Training Binary Weight Network only. 
`th main.lua -data [path to ImageNet dataset] -nGPU 1 -batchSize 128 -netType alexnet -binaryWeight -dropout 0.1`
Do you have similar options or can you tell me which parts need to be changed to do that?

Thanks!

**Comment by jiecaoyu** [comment]
Hi @lusensama , try commenting out [this line](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/ebccb503b2f9ebc96187a75b6daa63d95b61d110/ImageNet/networks/model_list/alexnet.py#L51). Let me know if it is not working.

**Comment by lusensama** [comment]
> Hi @lusensama , try commenting out [this line](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/ebccb503b2f9ebc96187a75b6daa63d95b61d110/ImageNet/networks/model_list/alexnet.py#L51). Let me know if it is not working.

Thank you!
That's what I tried! While I'm waiting for the training result, do you know if there's any way to speed up the training? it seems to be training much slower than the original XNOR-Net code. Is it because of the data-loading process of the pytorch dataloader?

Thanks!

**Comment by jiecaoyu** [comment]
@lusensama You can try to use [accimage](https://github.com/pytorch/accimage) to accelerate the data loading.

**Comment by lusensama** [comment]
> @lusensama You can try to use [accimage](https://github.com/pytorch/accimage) to accelerate the data loading.

Isn't that already part of [your code](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/master/ImageNet/networks/main.py#L168)?

**Comment by jiecaoyu** [comment]
@lusensama Okay, I thought you may disable this line as others do. If accimage has already been used, the data loading should not be the bottleneck. You can check the log file and it should list the average data loading time. How much slower is training? Are you comparing the training time with my log file or your own running?

**Comment by lusensama** [comment]
Here's a part of the log that I took randomly:
`Epoch: [2][4410/5005]   Time 0.159 (2.386)      Data 0.079 (2.300)      Loss 5.0986 (5.0895)    Prec@1 7.031 (9.000)    Prec@5 20.703 (23.295)
Epoch: [2][4420/5005]   Time 15.926 (2.389)     Data 15.841 (2.303)     Loss 5.0091 (5.0894)    Prec@1 9.375 (8.998)    Prec@5 25.781 (23.296)
Epoch: [2][4430/5005]   Time 0.160 (2.388)      Data 0.079 (2.302)      Loss 5.0597 (5.0894)    Prec@1 8.984 (8.996)    Prec@5 22.656 (23.295)
Epoch: [2][4440/5005]   Time 2.120 (2.388)      Data 1.950 (2.302)      Loss 5.0584 (5.0893)    Prec@1 8.594 (8.996)    Prec@5 19.141 (23.296)
Epoch: [2][4450/5005]   Time 0.195 (2.387)      Data 0.090 (2.301)      Loss 5.0794 (5.0896)    Prec@1 11.719 (8.994)   Prec@5 25.000 (23.294)
Epoch: [2][4460/5005]   Time 16.806 (2.390)     Data 16.704 (2.304)     Loss 5.0873 (5.0899)    Prec@1 6.641 (8.990)    Prec@5 23.047 (23.289)
Epoch: [2][4470/5005]   Time 0.187 (2.389)      Data 0.106 (2.303)      Loss 5.5341 (5.0904)    Prec@1 6.250 (8.986)    Prec@5 18.359 (23.282)
Epoch: [2][4480/5005]   Time 1.480 (2.389)      Data 1.399 (2.303)      Loss 5.0857 (5.0907)    Prec@1 10.156 (8.985)   Prec@5 24.609 (23.280)
Epoch: [2][4490/5005]   Time 0.161 (2.388)      Data 0.081 (2.302)      Loss 5.1710 (5.0910)    Prec@1 8.203 (8.982)    Prec@5 24.219 (23.274)`
which is much slower as compared to your baseline time on average. Although I have added a few more dropout layers, I think that shouldn't affect the training time so much...

I have two GPUs of Titan-X on ubuntu just FYI.

**Comment by jiecaoyu** [comment]
It seems the data loading becomes the bottleneck. You can try disabling accimage to see what happens.

**Comment by lusensama** [comment]
Epoch: [0][0/5005]      Time 31.310 (31.310)    Data 15.462 (15.462)    Loss 6.9078 (6.9078)    Prec@1 0.391 (0.391)    Prec@5 0.391 (0.391)
Epoch: [0][10/5005]     Time 0.362 (3.170)      Data 0.282 (1.646)      Loss 6.9079 (6.9078)    Prec@1 0.391 (0.071)    Prec@5 0.391 (0.355)
Epoch: [0][20/5005]     Time 0.178 (2.282)      Data 0.099 (1.443)      Loss 6.9143 (6.9079)    Prec@1 0.000 (0.056)    Prec@5 0.000 (0.298)
Epoch: [0][30/5005]     Time 0.173 (2.002)      Data 0.094 (1.405)      Loss 6.9051 (6.9072)    Prec@1 0.000 (0.063)    Prec@5 0.391 (0.315)
Epoch: [0][40/5005]     Time 11.687 (2.158)     Data 11.568 (1.684)     Loss 6.9085 (6.9065)    Prec@1 0.000 (0.086)    Prec@5 0.000 (0.391)
Epoch: [0][50/5005]     Time 0.252 (2.028)      Data 0.134 (1.627)      Loss 6.9064 (6.9058)    Prec@1 0.000 (0.092)    Prec@5 0.391 (0.414)
Epoch: [0][60/5005]     Time 0.201 (1.926)      Data 0.077 (1.576)      Loss 6.8913 (6.9049)    Prec@1 0.000 (0.096)    Prec@5 0.391 (0.442)
Epoch: [0][70/5005]     Time 0.156 (1.856)      Data 0.077 (1.544)      Loss 6.9021 (6.9042)    Prec@1 0.000 (0.094)    Prec@5 0.391 (0.446)
Epoch: [0][80/5005]     Time 9.707 (1.925)      Data 9.628 (1.641)      Loss 6.8964 (6.9034)    Prec@1 0.391 (0.096)    Prec@5 0.391 (0.434)
Epoch: [0][90/5005]     Time 2.907 (1.900)      Data 2.827 (1.638)      Loss 6.8834 (6.9019)    Prec@1 0.781 (0.116)    Prec@5 0.781 (0.455)
Epoch: [0][100/5005]    Time 0.164 (1.853)      Data 0.084 (1.609)      Loss 6.8882 (6.8996)    Prec@1 0.000 (0.112)    Prec@5 0.000 (0.445)
Epoch: [0][110/5005]    Time 0.155 (1.812)      Data 0.075 (1.582)      Loss 6.8952 (6.8997)    Prec@1 0.000 (0.102)    Prec@5 0.000 (0.447)
Epoch: [0][120/5005]    Time 8.798 (1.848)      Data 8.719 (1.630)      Loss 6.9373 (6.8980)    Prec@1 0.000 (0.097)    Prec@5 0.391 (0.449)
Epoch: [0][130/5005]    Time 3.426 (1.846)      Data 3.300 (1.638)      Loss 6.9039 (6.8978)    Prec@1 0.000 (0.092)    Prec@5 0.781 (0.453)

It seems to have no effect for disabling that; and it looks like it is experiencing a very slow loading every few batches.

**Comment by jiecaoyu** [comment]
Looks like disabling accimage makes things a little bit faster. Sorry but I am not sure what is happening there.

**Comment by lusensama** [comment]
> Looks like disabling accimage makes things a little bit faster. Sorry but I am not sure what is happening there.

It's alright! I will sort it out myself!
Thank you!
